### PR TITLE
fix(worker): mock DNS in url-normalization tests to remove CI flakiness

### PR DESCRIPTION
**Pull request by LysanderKie**
🟡 **Live preview: [pr-16128.preview.langfuse.com](https://pr-16128.preview.langfuse.com)**

## Summary

`worker/src/__tests__/url-normalization.test.ts` made real DNS lookups (via `validateWebhookURL` -> `resolveHost`) for hostnames like a 50-level subdomain and a bare `metadata` label. The latter is subject to resolver search-domain suffixing, which hung past the vitest timeout on CI and failed three tests after 3 retries each.

The first attempt at this copied the `vi.mock("node:dns/promises")` scaffolding from `webhook-validation.test.ts`, and it did not work: CI on that commit failed the same three tests. `vi.mock` only intercepts modules vitest loads, but this file reaches `validateWebhookURL` through `@langfuse/shared/src/server`, which the package `exports` map resolves to the compiled CommonJS output — its `require("node:dns/promises")` kept using the real resolver. The two suites that already mocked DNS work only because they import the shared TypeScript source by relative path.

So this stubs the `dns` module object with `vi.spyOn` instead. Node builtins are a single shared object, so the stubs apply through the compiled output too, and the source-vs-package import distinction stops being load-bearing. Applying the same shape to `webhook-validation.test.ts` and `llm-base-url-validation.test.ts` also drops the triplicated `vi.hoisted`/`vi.mock` block those suites carried.

Two follow-on changes:
- The bare `metadata` label now resolves to `169.254.169.254` in the stub, the way it does inside a cloud VM. That test asserts `Blocked (hostname|IP address) detected` instead of also accepting `DNS lookup failed`, so it covers the resolved-IP block it is named after — and every suite here now fails loudly if the stubs stop applying.
- Dropped the ad hoc `10000` ms per-test timeout, which was a workaround for the same flakiness.

## Test plan

- [x] `pnpm --filter worker run test src/__tests__/{url-normalization,webhook-validation,llm-base-url-validation}.test.ts` — `Tests  64 passed (64)`, and the suites run in 19ms total vs 226ms before, which is the observable proof the stubs apply (no resolver round-trips)
- [x] `pnpm run lint` — `Tasks: 7 successful, 7 total`
- [x] `pnpm run typecheck` — `Tasks: 7 successful, 7 total`

Tests added: 0 — test-infrastructure only; no new behavior to protect. The metadata case is an existing test whose assertion got tightened.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

<details open><summary><h3>Greptile Summary</h3></summary>

The PR makes URL-normalization tests network-independent by mocking all promise-based DNS APIs to return ENOTFOUND and removes the now-unnecessary extended timeout.
- Adds hoisted resolver mocks reset before every test.
- Replaces live DNS behavior with deterministic lookup failures.
- Restores the many-subdomains test to the default timeout.
</details>

<details><summary><h3>Confidence Score: 4/5</h3></summary>

The PR appears safe to merge, though the unconditional DNS failure weakens several tests and the import placement should be cleaned up.

Production behavior is unchanged, but the new test setup prevents the bare metadata and hostname-boundary cases from exercising the behavior their test names imply.

**Files Needing Attention:** worker/src/__tests__/url-normalization.test.ts
</details>

<details><summary>Prompt To Fix All With AI</summary>

`````markdown
### Issue 1
worker/src/__tests__/url-normalization.test.ts:38-46
**DNS failures bypass tested behavior**

The unconditional `ENOTFOUND` responses make the bare `metadata`, 253-character hostname, and 50-level subdomain cases terminate at the generic DNS-failure path. Those tests therefore pass without exercising resolved-IP blocking or the hostname boundary behavior they claim to cover, leaving regressions in those paths undetected.

### Issue 2
worker/src/__tests__/url-normalization.test.ts:29
**Import follows mock setup**

The `@langfuse/shared/src/server` import now appears after the hoisted mock declaration. Keep imports at the top of the module so dependencies remain immediately discoverable and the file follows the repository's required import organization.

---

For each issue above, determine whether it is valid and should be fixed. If so, fix it directly.
`````

</details>

<sub>Reviews (1): Last reviewed commit: ["fix(worker): mock DNS in url-normalizati..."](https://github.com/langfuse/langfuse/commit/f964a27af876c1a19f5ffd77c386552824bd9b37) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=53294283)</sub>

**Context used:**

- Rule used - Move imports to the top of the module instead of p... ([source](https://app.greptile.com/personal-org-4986/-/custom-context?memory=c960fc07-9928-409f-a18b-a780cbdded12))

**Learned From**
[langfuse/langfuse-python#1387](https://github.com/langfuse/langfuse-python/pull/1387)

### CODE REVIEW

File: worker/src/__tests__/url-normalization.test.ts [hunk 1]
```diff
@@ -1,6 +1,51 @@
...
+    lookup: lookupMock,
+  },
+  resolve4: resolve4Mock,
+  resolve6: resolve6Mock,
+  lookup: lookupMock,
+}));
+
 import { validateWebhookURL } from "@langfuse/shared/src/server";
 
+const dnsError = (code: string, hostname: string) =>
```
[the remark this answers is not in the document]
Comment (LysanderKie) [reply worker/src/__tests__/url-normalization.test.ts]: Valid, and fixed — though not by extracting a helper.

The three copies were not byte-for-byte equivalent, and the difference was the bug: `webhook-validation.test.ts` and `llm-base-url-validation.test.ts` import the shared TypeScript source by relative path, while this file imported `@langfuse/shared/src/server`, i.e. the compiled CommonJS output. `vi.mock` does not intercept its `require("node:dns/promises")`, so this copy was inert and CI on the previous commit failed the same three tests it was meant to fix. The failure mode you predicted ("a missed file means one suite silently reverts to real DNS lookups") had already happened in the file under review.

A shared `dnsMock.ts` would not have caught that, and as you note the `vi.mock` registration would still be per-file. So all three suites now use `vi.spyOn(dns, ...)` on the `node:dns/promises` module object: ~6 lines instead of ~30, no hoisting rules, and — because Node builtins are one shared object — the stubs apply through the compiled output too, which removes the source-vs-package import trap entirely.

The metadata test also now models a bare `metadata` label resolving to `169.254.169.254` and asserts `Blocked (hostname|IP address) detected`, so an inert stub fails the suite instead of passing it.

**Comment by LysanderKie** [comment]
Pushed 16a5078 — the previous commit did not work, and the review round missed it.

**What was wrong:** `vi.mock("node:dns/promises")` only intercepts modules vitest loads. This file imports `@langfuse/shared/src/server`, which the package `exports` map resolves to compiled CommonJS; its `require("node:dns/promises")` kept hitting the real resolver. CI on `f964a27` failed the same three tests at `url-normalization.test.ts:302`, `Test timed out in 5000ms`, 20s each with `retries=3`. Locally the suite ran in 226ms (real lookups) and now runs in 6ms.

**Fix:** `vi.spyOn(dns, ...)` on the module object, applied to all three DNS-mocking suites. Node builtins are a single shared object, so it works through the compiled output as well.

**On the Greptile issues:**

- *Issue 1 — DNS failures bypass tested behavior:* partly right, wrong cases. In the metadata test, `169.254.169.254`, `[fd00:ec2::254]` and `metadata.google.internal` are rejected by `isHostnameBlocked`/`isIPBlocked` before any DNS call (`outbound-url/validation.ts:127-143`), so blanket `ENOTFOUND` never touched them. Only the bare `metadata` case was weakened — fixed by resolving it to `169.254.169.254` and tightening the assertion to `Blocked (hostname|IP address) detected`. The 253-character and 50-subdomain cases were always plain DNS-failure assertions, not hostname-boundary coverage.
- *Issue 2 — import follows mock setup:* moot now, the file has no hoisted block left. It was also not a defect: `vi.mock` factories have to precede the import of the module under test, and the ordering matched both sibling suites.

**Calibration note for the bots:** both reviews validated the mock by reading it against `resolveHost`'s call sites — "the mock resolves synchronously instead of hitting real DNS", 4/5 confidence — while the PR's own required checks were red on exactly the tests the PR claimed to fix. Worth checking CI status on the head commit before concluding that a flakiness fix works.

**Comment by LysanderKie** [comment]
@claude review

**Comment by LysanderKie** [comment]
@claude review
